### PR TITLE
ramips: add support for 6.6 kernel version

### PR DESCRIPTION
**Pull request by namiltd**
Initial support for 6.6 kernel version.

Build system: x86/64
Build-tested: ramips/mt7620/tplink_archer-c5-v4
Run-tested: ramips/mt7620/tplink_archer-c5-v4

obraz

### CODE REVIEW

**Comment by msandber** [comment]
Building fails with message:
```
make[2]: Leaving directory '/home/ubuntu/git/openwrt/scripts/config'
/home/ubuntu/git/openwrt/include/kernel-version.mk:11: *** Missing kernel version/hash file for 6.6. Please create /home/ubuntu/git/openwrt/include/kernel-6.6.  Stop.
time: target/linux/prereq#0.01#0.00#0.01
    ERROR: target/linux failed to build.
```

Rebasing to https://github.com/openwrt/openwrt/pull/14751 fixes it.

**Comment by namiltd** [comment]
@msandber 
```
git clone https://github.com/hackpascal/openwrt-dev.git
cd openwrt-dev
git checkout linux-v6.6
wget https://github.com/openwrt/openwrt/pull/14774.patch -O 00.patch
git apply 00.patch -v
```
Target System -> Mediatek Ralink Mips
Subtarget->MT7620 Based Boards
Target Profile ->TP-Llink Archer C5 v4
Luci->Collections->luci *
use testing kernel

**Comment by namiltd** [comment]
Compilation goes smoothly but unfortunately, the compiled system does not start and this requires a more detailed investigation. I only have:
```
[    0.000000] Linux  6.6.20 (builder@buildhost) (mipsel-openwrt-linux-musl-gcc (OpenWrt GCC 12.3.0 r25430-eb44f39e39) 12.3.0, GNU ld (GNU Binutils) 2.40.0) Mar 6 11:40:57 2024
[    0.000000] Board has DDR2
[    0.000000] Analog PMU set to hw control
[    0.000000] Digital PMU set to hw control
[    0.000000] SoC Type: MediaTek MT7620A ver:2 eco:6
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019650 (MIPS 24KEc)
[    0.000000] MIPS: machine is TP-Link Archer C5 v4
[    0.000000] detected buffer overflow in memcmp
```
It's probably a memcmp error when calling detect_memory_region which shouldn't be displayed.

**Comment by DragonBluep** [comment]
> Compilation goes smoothly but unfortunately the compiled system does not start and this requires a more detailed investigation. I only have:
> 
> ```
> [    0.000000] Linux  6.6.20 (builder@buildhost) (mipsel-openwrt-linux-musl-gcc (OpenWrt GCC 12.3.0 r25430-eb44f39e39) 12.3.0, GNU ld (GNU Binutils) 2.40.0) Mar 6 11:40:57 2024
> [    0.000000] Board has DDR2
> [    0.000000] Analog PMU set to hw control
> [    0.000000] Digital PMU set to hw control
> [    0.000000] SoC Type: MediaTek MT7620A ver:2 eco:6
> [    0.000000] printk: bootconsole [early0] enabled
> [    0.000000] CPU0 revision is: 00019650 (MIPS 24KEc)
> [    0.000000] MIPS: machine is TP-Link Archer C5 v4
> [    0.000000] detected buffer overflow in memcmp
> ```
> 
> It's probably a memcmp error when calling detect_memory_region which shouldn't be displayed.

Reverting https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/include/linux/fortify-string.h?h=next-20240312&id=439a1bcac648fe9b59210cde8991fb2acf37bdab can workaround this issue. Not sure if GCC13 has some help.

**Comment by msandber** [comment]
> Compilation goes smoothly but unfortunately the compiled system does not start and this requires a more detailed investigation. I only have:

A mt7621 build , which is based on top of this PR,  boots without problems all the way. If that is of any help narrowing down the cause.

```
[    0.000000] Linux version 6.6.20 (ubuntu@ubuntu) (mipsel-openwrt-linux-musl-gcc (OpenWrt GCC 12.3.0 r25155-35f4442275) 12.3.0, GNU ld (GNU Binutils) 2.40.0) #0 SMP Tue Mar 12 05:27:34 2024
[    0.000000] SoC Type: MediaTek MT7621 ver:1 eco:3
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 0001992f (MIPS 1004Kc)
[    0.000000] MIPS: machine is DNA Valokuitu Plus EX400
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] VPE topology {2} total 2
[    0.000000] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
...
```

**Comment by msandber** [comment]
@namiltd And oh yes, kernel 6.6 PR was merged yesterday in the master if you want to rebase to it. :)

**Comment by PolynomialDivision** [comment]
Please bring the PR in a state that allows to review it. Split the commits like its done on all other kernel update PRs. Thanks! :)

**Comment by timocapa** [comment]
> > Compilation goes smoothly but unfortunately the compiled system does not start and this requires a more detailed investigation. I only have:
> 
> A mt7621 build , which is based on top of this PR, boots without problems all the way. If that is of any help narrowing down the cause.
> 
Made a mt7621 build, with your changes rebased on top of main, enabling “Strip unnecessary exports from the kernel image” seems to cause a build failure.

Didn’t want to spam this PR so I’m replying to you here. Not sure if this was introduced with 6.6 in general, or by one of the changes here. 

If you haven’t checked by the time I get home, I’ll attach a build log here. :)

**Comment by namiltd** [comment]
After using https://github.com/openwrt/openwrt/pull/14774/commits/35ca3a008bef003c1f139b48b6f88cff39c2afa8 the problem with "buffer overflow in memcmp" disappeared. The router is up normally, but the 2.4GHz network does not work, the network is visible but I cannot connect to it.

**Comment by graysky2** [comment]
Great job with this PR, @namiltd.  Regarding https://github.com/openwrt/openwrt/commit/26ee2f88bb713afc586d822b980beef4086b2f8f

Build system: x86/64
Build-tested: ramips/tplink_archer-a6-v3
Run-tested: ramips/tplink_archer-a6-v3

I am not using the wireless of this device/it is basically a smart switch but I see no dmesg regressions/everything works as expected.  Tried both 6.6.20 and after merging https://github.com/openwrt/openwrt/pull/14909 to 6.6.22.

**Comment by castillofrancodamian** [comment]
@graysky2 Snapshots for this target have not been compiled for several days.

**Comment by namiltd** [comment]
I think the code is complete now and ready for commit.

**Comment by the2masters** [comment]
I tested this on a Ubiquiti U6-LR, it's working fine for 20 minutes now and I could not spot any new warnings / errors.

Tested-by: Stefan Hellermann <stefan@the2masters.de>

**Comment by namiltd** [comment]
I think everything is done and ready to merge.

**Comment by timocapa** [comment]
Build system: x86/64
Build-tested: ramips/mt7621/xiaomi_mi-router-3g / xiaomi_redmi-router-ac2100
Run-tested: ramips/mt7621/xiaomi_mi-router-3g / xiaomi_redmi-router-ac2100

Rebased this PR on top of main, I see no issues.

**Comment by PolynomialDivision** [comment]
Can you please adjust the commit message? You have 3 times "Fix compilation error on kernel 6.6". Be a bit more specific. Further, the patch from @DragonBluep is put into generic, so it is not ath79 specific. Also we don't write "core_packages", instead we use "kernel: modules: netdevices:".

**Comment by namiltd** [comment]
@PolynomialDivision Done

**Comment by namiltd** [comment]
I didn't change the code but:
[Build all core packages for selected target / Build Packages with external toolchain (malta, be) / Build malta/be](https://github.com/openwrt/openwrt/actions/runs/8400945171/job/23009460524?pr=14774#logs)
and [Build all core packages for selected target / Build Packages with external toolchain (x86, 64) / Build x86/64](https://github.com/openwrt/openwrt/actions/runs/8400945171/job/23009460910?pr=14774#logs) fails.
Problem with buildbots?

**Comment by LynxChaus** [comment]
Attempt to boot UBNT EdgeRouter X SFP lead to kernel crash in at8031_probe():

```
[    4.629622] mt7530-mdio mdio-bus:1f: MT7530 adapts as multi-chip module
[    4.658499] CPU 1 Unable to handle kernel paging request at virtual address 0000000b, epc == 80523f6c, ra == 80523f40
[    4.679711] Oops[#1]:
[    4.684233] CPU: 1 PID: 1 Comm: swapper/0 Not tainted 6.6.22 #0
[    4.696007] $ 0   : 00000000 00000001 00000001 00000001
[    4.706423] $ 4   : 81ea4560 8184a580 00000122 00000000
[    4.716834] $ 8   : 00000001 807a553c 00000000 f00635a6
[    4.727246] $12   : 3da73a4e 29f7fb3a b7920b13 9430ed3b
[    4.737657] $16   : 8190f400 00000001 00000000 00000000
[    4.748068] $20   : 00000000 80ff1038 808f4b80 809e0000
[    4.758480] $24   : 00000000 80558250                  
[    4.768891] $28   : 81836000 81837928 80ff0104 80523f40
[    4.779304] Hi    : 00000009
[    4.785023] Lo    : 2a202c20
[    4.790741] epc   : 80523f6c at8031_probe+0x150/0x204
[    4.800810] ra    : 80523f40 at8031_probe+0x124/0x204
[    4.810863] Status: 1100fc03 KERNEL EXL IE 
[    4.819193] Cause : 0080000c (ExcCode 03)
[    4.827162] BadVA : 0000000b
[    4.832879] PrId  : 0001992f (MIPS 1004Kc)
[    4.841016] Modules linked in:
[    4.847086] Process swapper/0 (pid: 1, threadinfo=(ptrval), task=(ptrval), tls=00000000)
[    4.863184] Stack : 81b35688 8027ba78 00000000 80417eb8 804180b8 80ff1038 8190f400 00000000
[    4.879842]         8190f400 00000000 00000000 00000000 8190f400 809e6004 809e6004 8051c598
[    4.896498]         00000000 00000004 81b32108 808b7ee8 81b35688 8027f9a0 80ff1038 804968e0
[    4.913153]         81ea522d 80ff1108 8190f400 00000000 809e6004 00000000 8190f400 00000000
[    4.929806]         809e6004 00000000 00000000 80ff1038 808f4b80 809e0000 80ff0104 804974d8
[    4.946462]         ...
[    4.951325] Call Trace:
[    4.956176] [<80523f6c>] at8031_probe+0x150/0x204
[    4.965544] [<8051c598>] phy_probe+0x8c/0x504
[    4.974237] [<804974d8>] really_probe+0xd4/0x368
[    4.983448] [<804979a0>] driver_probe_device+0x4c/0x154
[    4.993853] [<80497b78>] __device_attach_driver+0xd0/0x15c
[    5.004775] [<804951f4>] bus_for_each_drv+0x7c/0xc8
[    5.014487] [<80498044>] __device_attach+0xf8/0x16c
[    5.024194] [<80496448>] bus_probe_device+0xb4/0xdc
[    5.033906] [<80493780>] device_add+0x578/0x7a4
[    5.042925] [<8051996c>] phy_device_register+0x5c/0xa8
[    5.053155] [<805240e8>] fwnode_mdiobus_phy_device_register+0xc8/0x16c
[    5.066150] [<80524344>] fwnode_mdiobus_register_phy+0x1b8/0x1d0
[    5.078106] [<80524554>] __of_mdiobus_parse_phys+0x12c/0x26c
[    5.089372] [<80524aec>] __of_mdiobus_register+0x100/0x26c
[    5.100292] [<80530bd4>] mtk_probe+0xf04/0x1118
[    5.109315] [<8049a088>] platform_probe+0x50/0xa4
[    5.118684] [<804974d8>] really_probe+0xd4/0x368
[    5.127877] [<804979a0>] driver_probe_device+0x4c/0x154
[    5.138280] [<80497cb4>] __driver_attach+0xb0/0x1c0
[    5.147992] [<80495130>] bus_for_each_dev+0x70/0xb8
[    5.157702] [<804966cc>] bus_add_driver+0x100/0x204
[    5.167413] [<804989a4>] driver_register+0x88/0x154
[    5.177119] [<80001644>] do_one_initcall+0x50/0x2e0
[    5.186827] [<809f211c>] kernel_init_freeable+0x284/0x314
[    5.197573] [<807a6044>] kernel_init+0x20/0x114
[    5.206595] [<80002b18>] ret_from_kernel_thread+0x14/0x1c
[    5.217345] 
[    5.220298] Code: 2e310002  12200004  24020001 <a242000b> 24020001  a242000a  02002025  afa00010  24070020 
[    5.239727] 
[    5.242734] ---[ end trace 0000000000000000 ]---
[    5.251937] Kernel panic - not syncing: Fatal exception
[    5.262341] Rebooting in 30 seconds..
``` 

objdump show code block moved in patch target/linux/generic/backport-6.6/707-v6.8-06-net-phy-at803x-move-specific-at8031-probe-mode-check.patch - priv dereference.
@Ansuel is patch correct?

**Comment by namiltd** [comment]
Take a look at: https://github.com/openwrt/openwrt/pull/14938
Edgerouter X does not have enough RAM for kernel 6.1 as well as 6.6.

**Comment by LynxChaus** [comment]
@namiltd - i see this commit. But this is "official" openwrt decision due to vendor u-boot and it's silly mtd partition. But - I'm use  original vendor u-boot ability to boot image from network for testing SFP module cage via at8031.

**Comment by PolynomialDivision** [comment]
Can you add more information? As a reviewer I would now need to look up everything by myself. It would speed up the reviewing process if you link to the upstream changes done in the linux kernel that makes the adaption necessary. Probably you already did that to figure out the necessary adaptions to the patches.

**Comment by namiltd** [comment]
I looked at the code about 2 weeks ago, so it will take some work. I understand that this information should be included in the descriptions of subsequent amendments? 
For example https://github.com/openwrt/openwrt/pull/14774/commits/17d2670c504464e1f0fa3d73aca6f1d0ea0dd566 follows https://github.com/torvalds/linux/commit/44b0c2957adc62b86fcd51adeaf8e993171b c319
Should I add this?

**Comment by Spudz76** [comment]
Build system: x86/64
Build-tested: ramips/mt7621/linksys_ea7300-v1
Run-tested: ramips/mt7621/linksys_ea7300-v1

`Linux wonky2.lan 6.6.22 #0 SMP Wed Mar 27 07:01:26 2024 mips GNU/Linux`

Applied on top of main using diff URL piped to `patch -p1`, Boots and looks normal, did not fully test all functionalities, but Ethernet works and Wireless came up.

`kmod-crypto-hw-eip93` also appears to work fine, tested and compared to 6.1 with `wolfssl-benchmark`

`kmod-ovpn-dco` failed to build but that's not really surprising, it barely ever works correctly with 6.1 either.

**Comment by namiltd** [comment]
The kmod-ovpn-dco module does not compile also on other platforms and requires individual checking and correction.

**Comment by namiltd** [comment]
@PolynomialDivision All done

**Comment by 9uyone** [comment]
After switching to 6.6, I encountered the following problem: in Windows 11, the network was not founding at all; Wi-Fi worked flawlessly though.
After googling, I tried to manually set the 100Mb full duplex mode in Windows control panel — it have worked.
It would be good to deal and fix it.

**mt7621**

**Comment by timocapa** [comment]
Might be device specific?

Both my RM2100 and R3G seem to auto negotiate 1 gig without issues and my devices connect just fine

**Comment by 9uyone** [comment]
And now, after few reboots and some attempts to connect directly with different cables, I was able to start on **auto negotiation** with the same cable. Maybe some bad contact played a role, lol

**Comment by Vladdrako** [comment]
Tested on mt7620/asus_rt-ac51u
A little typo :)
```[    0.689188] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]```

File: target/linux/ramips/mt7620/config-6.6 [hunk 1]
```diff
@@ -167,7 +173,8 @@ CONFIG_PHY_RALINK_USB=y
 CONFIG_PINCTRL=y
 # CONFIG_PINCTRL_AW9523 is not set
 CONFIG_PINCTRL_MT7620=y
-CONFIG_PINCTRL_RALINK=y
+CONFIG_PINCTRL_MT76X8=y
```
Comment (DragonBluep) [inline target/linux/ramips/mt7620/config-6.6:176]: `CONFIG_PINCTRL_MT76X8` driver is useless.

Comment (DragonBluep) [inline target/linux/ramips/mt7620/config-6.6:176]: `CONFIG_PINCTRL_MT76X8` driver is useless.

File: target/linux/ramips/mt7620/config-6.6 [hunk 2]
```diff
@@ -191,12 +199,11 @@ CONFIG_SPI_MASTER=y
 CONFIG_SPI_MEM=y
 # CONFIG_SPI_MT7621 is not set
 CONFIG_SPI_RT2880=y
-CONFIG_SRCU=y
+CONFIG_SQUASHFS_DECOMP_MULTI_PERCPU=y
```
Comment (DragonBluep) [inline target/linux/ramips/mt7620/config-6.6:202]: Multi thread is useless for single core CPU. Please also disable it for other non MT7621 targets.

File: target/linux/ramips/mt76x8/config-6.6 [hunk 3]
```diff
@@ -161,7 +167,8 @@ CONFIG_PHY_RALINK_USB=y
 CONFIG_PINCTRL=y
 # CONFIG_PINCTRL_AW9523 is not set
 CONFIG_PINCTRL_MT7620=y
```
Comment (DragonBluep) [inline target/linux/ramips/mt76x8/config-6.6:169]: `CONFIG_PINCTRL_MT7620` driver is useless.

**Comment by namiltd** [comment]
@DragonBluep All done

**Comment by Spudz76** [comment]
> `[ 0.689188] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]`

[not a typo](https://elixir.bootlin.com/linux/latest/source/include/linux/pci.h#L657)

It's short for Bus Number (busn) Resource

File: target/linux/ramips/mt7620/config-6.6 [hunk 4]
```diff
@@ -191,12 +199,11 @@ CONFIG_SPI_MASTER=y
 CONFIG_SPI_MEM=y
 # CONFIG_SPI_MT7621 is not set
 CONFIG_SPI_RT2880=y
-CONFIG_SRCU=y
+# CONFIG_SQUASHFS_DECOMP_MULTI_PERCPU is not set
```
Comment (DragonBluep) [inline target/linux/ramips/mt7620/config-6.6:202]: It seems that some symbols are missing. Please always use `make kernel_menuconfig` to refresh the config file.

**Comment by namiltd** [comment]
@DragonBluep All done

**Comment by namiltd** [comment]
Is there a chance for approval before the kernel is updated to 6.6.24 (released today) and we have to adapt the patches again?

**Review by DragonBluep — APPROVED** [review]
> @DragonBluep Is there a chance for approval before the kernel is updated to 6.6.24 (released today) and we have to adapt the patches again?

LGTM. `Reviewed-by: Shiji Yang <yangshiji66@outlook.com>`.

Thank you for your contribution. I am not a repo maintainer, so I cannot provide further help.

File: target/linux/generic/pending-6.6/350-mips-kernel-fix-detect_memory_region-function.patch
Comment (robimarko) [inline target/linux/generic/pending-6.6/350-mips-kernel-fix-detect_memory_region-function.patch:1]: Has this been sent upstream?

Comment (namiltd) [reply target/linux/generic/pending-6.6/350-mips-kernel-fix-detect_memory_region-function.patch:1]: nope

Comment (robimarko) [reply target/linux/generic/pending-6.6/350-mips-kernel-fix-detect_memory_region-function.patch:1]: Great, it was sent

**Comment by PolynomialDivision** [comment]
Why is the CI failing?

**Comment by namiltd** [comment]
The bug does not affect this PR but other commits that have not been adapted to the current code version (bcm53xx and ipq40xx)

**Comment by PolynomialDivision** [comment]
Thanks. Merged to main! :)

**Pull request merged**